### PR TITLE
feat: re-organize ec-gpu traits

### PR DESCRIPTION
**Pull request by vmx**
The ec-gpu traits are re-organized.

BREAKING CHANGE: The `GpuEngine` trait is removed, as it was specific
to pairing friendlt elliptic curves.

The `GpuField` trait gains a new method called `sub_field_name`, which
only matters for extension fields. It has a default implementation, so
existing implementations don't need to change.

A new trait called `GpuName` is introduced, which every `GpuField` needs
to implement. It contains the name of the object it's implemented for,
which is then used in the generated source code as identifier.

### CODE REVIEW

**Comment by vmx** [comment]
@dignifiedquire I'd like to get a review from you as you did the original work on this crate.

@DrPeterVanNostrand I'd like especially a review of the comments.

For both of you: the implementation of the traits for `blstrs` can be found here: https://github.com/filecoin-project/blstrs/pull/43

File: ec-gpu/src/lib.rs
Comment (str4d) [inline ec-gpu/src/lib.rs:9]: This global uniqueness requirement is going to be very hard to enforce; in fact, this documentation recommends a strategy for naming fields that is guaranteed to cause collisions across different implementations of the same field.

However, we know that within a particular binary application, each type will have a unique [`TypeId`](https://doc.rust-lang.org/std/any/struct.TypeId.html). Would it be possible to generate the GPU source code from something based on that identifier, as derived from `Self`? The `TypeId` is allowed to change between compilation runs (see [this issue](https://github.com/rust-lang/rust/issues/61553) for some context), so it wouldn't work if the expectation is that the generated GPU source code can be compiled once and reused indefinitely, but if the source is being generated and compiled live (and thus is already subject to the `rustc -Cmetadata` flag) then this should both work, and mean downstream crates don't need to do anything (as the changes to `GpuField` could then be entirely non-breaking, either with this being a default method on `GpuField` or by inverting the trait bounds so `GpuName: GpuField` and providing a blanket implementation of `GpuName`).

Comment (vmx) [reply ec-gpu/src/lib.rs:9]: I thought about using `TypeId`, but then thought it might be too brittle. Practically I wouldn't expect that there are bazillions of implementations. I even thought one might perhaps even use a non-unique name intentionally. E.g. you might even want to create the same source for differently implemented fields (thought I'm not really sure if that's a use case). I liked the idea of keeping things pretty straight forward and easy to debug. Hence I think explicit names are better.

Input from other folks would be appreciated.

Even if we would use `TypeId` it would still be a breaking change for downstream crates, as there are cases where `GpuName` is implemented for things that aren't fields. So the trait bound cannot be flipped.

Comment (str4d) [reply ec-gpu/src/lib.rs:9]: > Practically I wouldn't expect that there are bazillions of implementations.

That may be true, but it only takes two to collide in order for the documented requirement to fail. And there are no clearly documented consequences for a collision.

> I even thought one might perhaps even use a non-unique name intentionally. E.g. you might even want to create the same source for differently implemented fields (thought I'm not really sure if that's a use case).

This is where it would be _very_ useful to have more documentation about how this function will (or should) be used. If the contents of this method are not being tightly constrained, then it falls to the consumer of the method to have a well-defined usage contract, so that people implementing this trait know what they are implementing it against.

> I liked the idea of keeping things pretty straight forward and easy to debug. Hence I think explicit names are better.

Definitely reasonable, and also why I'd prefer human-meaningful names over compiler-generated identifiers. If debuggability were the only concern, then [`std::any::type_name`](https://doc.rust-lang.org/std/any/fn.type_name.html) would be perfect for this; sadly it doesn't guarantee to include the module path, so it doesn't claim uniqueness.

Maybe the recommended implementation for this method should be:
```rust
    fn name() -> String {
        format!("{}::{}", module_path!(), std::any::type_name::<Self>())
    }
```
(or something more involved that strips the module path if present from the output of `type_name`). This could even be provided as a macro (used like e.g. `ec_gpu::name!(MyType);`) so that most people use the non-colliding default (while people with more convoluted use cases can implement the trait directly).

> Even if we would use `TypeId` it would still be a breaking change for downstream crates, as there are cases where `GpuName` is implemented for things that aren't fields. So the trait bound cannot be flipped.

Good to know.

Comment (vmx) [reply ec-gpu/src/lib.rs]: Thanks @str4d for all the input. I played a bit around with those ideas and pushed a new commit. Now the name is auto-generated and it actually isn't too bad. In case of BLS381-12, it's names like `ec_gpu_blstrs__fp__Fp`, `ec_gpu_blstrs__scalar__Scalar` or `ec_gpu_blstrs__g1__G1Affine`, which are still reasonable. I've also updated the documentation that you shouldn't implement the name yourself.

I thought it would be a good idea to auto-implement it for the `GpuField`. I don't know if it's still a breaking change (I guess it still is), but at least it will need less code changes. I've also updated the `blstrs` PR so that folks can see what changes would be needed https://github.com/filecoin-project/blstrs/pull/43. I think the `pasta_curves` trait wouldn't need any code changes.

Comment (vmx) [reply ec-gpu/src/lib.rs]: I forgot to add that `G1` and `G2` (https://github.com/filecoin-project/blstrs/pull/43/files#diff-a81fd3c080fd012c0c9155e68a28c5ef92bafcd33eced7847bbb40c5a7dcc116R841) are an example where I need a `GpuName` which aren't fields. I need it in order to name the Multiexp/MSM kernel correctly.

File: ec-gpu/src/lib.rs
Comment (dignifiedquire) [inline ec-gpu/src/lib.rs:4]: why make it its own trait and not part pf GpuField?

Comment (vmx) [reply ec-gpu/src/lib.rs:4]: See https://github.com/filecoin-project/ec-gpu/pull/30#discussion_r937057634, it's also used for things that are not fields.

File: ec-gpu/src/lib.rs
Comment (str4d) [inline ec-gpu/src/lib.rs:19]: This default implementation doesn't quite do the right thing. `module_path!()` inserts the path for the module in which the macro is evaluated, which in this case is `ec_gpu`. That's handy for domain-separating multiple versions of `ec-gpu`, but this could also just be done at the point where `GpuName::name()` gets called.

What I'd suggested it for was domain-separating the downstream types themselves, and that only works if the `module_path!()` macro is invoked inside the downstream modules. This is why in my previous comment I suggested providing a wrapping macro that downstreams could use to implement `GpuName`, which _would_ invoke it in the correct context.

The reason the examples you showed before look right is because `std::any::type_name` happens to be printing the module path, but it is not guaranteed to do so (it may also just show the type name). I think the best way to combine them would be to check if the output of `type_name()` is prefixed by `module_path!()` and prepend the latter if not. In reusable macro form:
```rust
#[macro_export]
macro_rules! name {
    ($named:ident) => {
        impl ::ec_gpu::GpuName for $named {
            fn name() -> String {
                let mod_path = module_path!();
                let type_name = std::any::type_name::<Self>();
                if type_name.starts_with(mod_path) {
                    type_name.replace(|c: char| !c.is_ascii_alphanumeric(), "_")
                } else {
                    format!("{}__{}", mod_path, type_name)
                        .replace(|c: char| !c.is_ascii_alphanumeric(), "_")
                }
            }
        }
    };
}
```
and then the default impl would be removed:
```suggestion
    fn name() -> String;
```

File: ec-gpu/src/lib.rs
Comment (str4d) [inline ec-gpu/src/lib.rs:43]: This would also be removed, as there'd not be a default impl:
```suggestion
```

File: ec-gpu/src/lib.rs
Comment (str4d) [reply ec-gpu/src/lib.rs:19]: The macro would then be used like so:
```rust
#[cfg(feature = "gpu")]
ec_gpu::name!(Fp2);

#[cfg(feature = "gpu")]
impl ec_gpu::GpuField for Fp2 {
    ...
}
```

**Comment by vmx** [comment]
I've pushed a new version, which I think is the best of the suggestions from @str4d and my idea. I found the macro a bit too magical. I prefer the code being more explicit and easy to grep. I updated the documentation, so that people know that they should use the `ec_gpu::name()` macro. @str4d what do you think?

File: ec-gpu/src/lib.rs
Comment (str4d) [inline ec-gpu/src/lib.rs:3]:
```suggestion
```

**Review by str4d — APPROVED** [review]
ACK. I confirmed that https://github.com/zcash/pasta_curves/pull/45 produces the names `pasta_curves__fields__fp__Fp` and `pasta_curves__fields__fq__Fq`.

**Comment by vmx** [comment]
I've squashed it into a single commit with an updated commit message.

@dignifiedquire it would be great if you could also review this code, as you are the maintainer of `blstrs`.

@DrPeterVanNostrand if you could have a look if I got the comments right, it would be great.

@str4d thanks once again for the thorough review, I'm really happy with the outcome.

**Review by dignifiedquire — APPROVED** [review]
lgtm

**Pull request merged**
